crates/memory/src/op_store/time_slice_hash_store.rs

```diff
@@ -0,0 +1,140 @@
+use kitsune2_api::{K2Error, K2Result};
+use std::collections::BTreeMap;
+
+/// In-memory store for time slice hashes.
+///
+/// Empty hashes do not need to be stored, and will be rejected with an error if you try to insert
+/// one. Hashes that are stored, are stored sparsely, and are indexed by the slice id.
+///
+/// It is valid to look up a time slice which has not had a hash stored, and you will get a `None`
+/// response. Otherwise, you will get exactly what was most recently stored for that slice id.
+#[derive(Debug, Default)]
+#[cfg_attr(test, derive(Clone))]
+pub(super) struct TimeSliceHashStore {
+    inner: BTreeMap<u64, bytes::Bytes>,
+}
+
+impl TimeSliceHashStore {
+    /// Insert a hash at the given slice id.
+    pub(super) fn insert(
```

Comment:
wow that is dramatically simpler haha. So I see that a BTreeMap gives us efficient compression of empty space as its simply not included. But if we have a bunch of sequential hashes, then in theory we don't need the index of any but the first. Do you know if serializing a BTreeMap takes this into account, or is it worth writing a custom serializer? Or is this an unnecessary premature optimization?

Reply:
I expect that sequences of non-empty hashes that are exactly equal will be incredibly rare. I don't think we need to worry about compressing those. I'd tried to deal with compressing empty hashes but would have been better off with this sparse representation that only includes hashes with some real content.

Even if you published the same entry in a Holochain app once a day and there was no other DHT activity, it'd still be at a different timestamp and so the op hash would be different. Just occasionally the XOR of a bunch of hashes will match, but not often enough that I think it's important to compress

Reply:
No sorry I don't mean sequential hashes that are equal, I mean that I assume the BTreeMap is always storing the index for every hash, even if the indexes are sequential.

I.e.  [..., (index, hash), (index+1, hash2), (index+2, hash3), ...] 

But because they're sequential you don't actually need index+1 and index+2, so I'm wondering if BTreeMap takes this into account in serialization or we should for a little more compression, or if you think it low value premature optimization.

Reply:
Oooh, I get what you're saying. When we do have a block of hashes because we've had data in each time slice.

Yes I think you could definitely write a data structure that does that. I don't think BTreeMap would.

However, this is really just a test implementation. We'll need a persistent version of this in SQLite and then we'd struggle to implement the same optimization. I'm quite happy to have the test implementation less than perfect on memory usage but really similar to the real implementation

Reply:
Gotcha makes sense. I'm mostly interested in optimizing the data sent over network rather than memory/disk usage.